presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineSemiJoinDistributionType.java

```diff
@@ -0,0 +1,167 @@
+    {
+        this.costComparator = requireNonNull(costComparator, "costComparator is null");
+        this.taskCountEstimator = requireNonNull(taskCountEstimator, "taskCountEstimator is null");
+    }
+
+    @Override
+    public Pattern<SemiJoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(SemiJoinNode semiJoinNode, Captures captures, Context context)
+    {
+        JoinDistributionType joinDistributionType = getJoinDistributionType(context.getSession());
+        if (joinDistributionType == AUTOMATIC) {
+            return Result.ofPlanNode(getCostBasedDistributionType(semiJoinNode, context));
+        }
+
+        else if (joinDistributionType.canPartition()) {
+            return Result.ofPlanNode(semiJoinNode.withDistributionType(PARTITIONED));
+        }
+
+        return Result.ofPlanNode(semiJoinNode.withDistributionType(REPLICATED));
+    }
+
+    private PlanNode getCostBasedDistributionType(SemiJoinNode node, Context context)
+    {
+        if (!canReplicate(node, context)) {
+            return node.withDistributionType(PARTITIONED);
+        }
+
+        List<PlanNodeWithCost> possibleJoinNodes = new ArrayList<>();
+        possibleJoinNodes.add(getSemiJoinNodeWithCost(node.withDistributionType(REPLICATED), context));
+        possibleJoinNodes.add(getSemiJoinNodeWithCost(node.withDistributionType(PARTITIONED), context));
+        
+        if (possibleJoinNodes.stream().anyMatch(result -> result.getCost().hasUnknownComponents()) || possibleJoinNodes.isEmpty()) {
+            return node.withDistributionType(PARTITIONED);
+        }
+
+        // Using Ordering to facilitate rule determinism
+        Ordering<PlanNodeWithCost> planNodeOrderings = costComparator.forSession(context.getSession()).onResultOf(PlanNodeWithCost::getCost);
+        return planNodeOrderings.min(possibleJoinNodes).getPlanNode();
+    }
+
+    private boolean canReplicate(SemiJoinNode node, Context context)
+    {
+        Optional<DataSize> joinMaxBroadcastTableSize = getJoinMaxBroadcastTableSize(context.getSession());
+        if (!joinMaxBroadcastTableSize.isPresent()) {
+            return true;
+        }
+
+        PlanNode buildSide = node.getFilteringSource();
+        PlanNodeStatsEstimate buildSideStatsEstimate = context.getStatsProvider().getStats(buildSide);
+        double buildSideSizeInBytes = buildSideStatsEstimate.getOutputSizeInBytes(buildSide.getOutputSymbols(), context.getSymbolAllocator().getTypes());
+        return buildSideSizeInBytes <= joinMaxBroadcastTableSize.get().toBytes();
+    }
+
+    private PlanNodeWithCost getSemiJoinNodeWithCost(SemiJoinNode possibleJoinNode, Context context)
+    {
+        TypeProvider types = context.getSymbolAllocator().getTypes();
+        StatsProvider stats = context.getStatsProvider();
+        boolean replicated = possibleJoinNode.getDistributionType().get().equals(REPLICATED);
+        /*
+         *   HACK!
```

Comment:
Can we remove the hack? (I don't mean to remove the word "HACK",  rather, the hack itself).

Reply:
I don't see how to do that.  See here for some previous discussion on the topic https://github.com/prestodb/presto/pull/11982#discussion_r237959979